fix issue 17940 - always set REX for 1-byte CSE moves on x86_64

FeepingCreature opened the pull request:
When generating a 1-byte mov in cse_simple on x86_64, we should always set rex,
not just when it's a register >= 4.
The reason is that our cse may be rewritten to mov into a reg >= 4 later.
This fixes https://issues.dlang.org/show_bug.cgi?id=17940 .

Thanks @belka-ew for helping minimize the testcase and testing the patch!

stefan-koch-sociomantic commented:
@FeepingCreature you need to reword the git commit.
`git rebase -i @~1`

WalterBright commented:
Can you please point in the generated code where the bad mov is?

FeepingCreature commented:
Sure.
```(dmd-2.076.1)feep@lop ~ $ dmd -O -betterC bug17940.d && obj2asm bug17940.o |tail -40 |head -16
                test    RBX,RBX
                je      L50
                test    RBX,RBX
                je      L2F
                movzx   EDI,byte ptr [RBX]
                call      _D8bug179404fun1FiZv@PLT32
L2F:            movsx   EDX,byte ptr [RBX]
                test    DL,DL
                je      L50
                mov     RCX,-8[RBP]
                mov     RAX,-010h[RBP]
                mov     RDX,RAX
                mov     DH,[RBX] <--- Here it is.
                and     ESI,0FFh <--- Note register.
                xor     EDI,EDI
                call      _D8bug179404fun2FS8bug179405ArrayiiZv@PLT32
```

WalterBright commented:
Thanks for tracking this down! I'd rather fix this at the register rewrite place. Delete lines 2263, 2264, and the same pair 24 lines up, in cod3.c.
Add these lines in cgcod.c after `cr->setReg(reg)`:
```
if (I64 && reg >= 4 && tysize(csextab[i].e->Ety) == 1)
    cr->Irex |= REX;
```
and see if that works!

FeepingCreature committed:
fix issue 17940
When generating a 1-byte mov cse on x86_64, handle the REX
flag after setReg, instead of when the instruction is first
created, since that's when we know if we need it.

FeepingCreature commented:
So done.

WalterBright commented:
Thanks!
